src/zkp/generator.rs

```diff
@@ -1,7 +1,115 @@
 use core::{fmt, str};
-use ffi;
-use {constants, from_hex, Error, Secp256k1, SecretKey, Signing, Tag};
+use ffi::{self, CPtr};
+#[cfg(feature = "rand")]
+use rand::Rng;
+use {constants, from_hex, Error, Secp256k1, Signing, Tag};
 
+/// Represents a blinding factor/Tweak on secp256k1 curve
+///
+/// Contrary to a [`crate::SecretKey`], the value 0 is also a valid tweak.
+/// Values outside secp curve order are invalid tweaks.
+pub struct Tweak([u8; constants::SECRET_KEY_SIZE]);
+impl_array_newtype!(Tweak, u8, constants::SECRET_KEY_SIZE);
+
+/// The zero Tweak
+pub const ZERO_TWEAK: Tweak = Tweak([
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+]);
+
+impl fmt::Debug for Tweak {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Tweak(")?;
+        for i in self[..].iter().cloned() {
+            write!(f, "{:02x}", i)?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl fmt::LowerHex for Tweak {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in &self.0[..] {
+            write!(f, "{:02x}", *ch)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for Tweak {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl str::FromStr for Tweak {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Tweak, Error> {
+        let mut res = [0; constants::SECRET_KEY_SIZE];
+        match from_hex(s, &mut res) {
+            Ok(constants::SECRET_KEY_SIZE) => Tweak::from_inner(res),
+            _ => Err(Error::InvalidTweakLength),
+        }
+    }
+}
+
+impl Tweak {
+    /// Generate a new random Tweak
+    #[cfg(feature = "rand")]
+    pub fn new<R: Rng + ?Sized>(rng: &mut R) -> Tweak {
+        let mut ret = [0u8; constants::SECRET_KEY_SIZE];
+        rng.fill_bytes(&mut ret);
+        Tweak(ret)
+    }
+
+    /// Converts a byte slice to a Tweak
+    /// Fails if tweak is not 32 bytes or if tweak is outside secp curve order
+    #[inline]
+    pub fn from_slice(data: &[u8]) -> Result<Tweak, Error> {
+        match data.len() {
+            constants::SECRET_KEY_SIZE => {
+                let mut ret = [0; constants::SECRET_KEY_SIZE];
+                unsafe {
+                    if ffi::secp256k1_ec_seckey_verify(
+                        ffi::secp256k1_context_no_precomp,
+                        data.as_ref().as_c_ptr(),
+                    ) == 0
+                    {
+                        if data.iter().all(|x| *x == 0) {
```

Comment:
I am not sure if this is a constant-time crypto concern. A special case of [0; 32].

Reply:
Here is my rationale for the above change. 
Assuming we care about constant time, we only care about constant time properties for valid keys. With the above code, we guarantee the constant time property for all valid tweaks but zero. 
Although this is not ideal, this suffices in elements use-case where zero is usually as explicit commitment. 

Checking this upfront might however leak information about valid Tweaks (based on the position of first 1). I can however move the code afterward with a short circuit return the valid non-zero tweak in the unsafe block. 

Or I may just be overthinking this :man_shrugging: .

Reply:
Thanks for clarifying, this does make sense! Happy to leave the code where it is.

Do you mind adding this information to the commit message?

Reply:
I think this is fine as-is.

Strictly speaking 0 is as good a tweak as any other value, and this is a violation of constant-timeness. Having said this, the only way to get 0 in practice is to do so deliberately -- "standard" use of CT means that these values will always be either uniformly random or forced by the balancing equation, and in both cases will be nonzero with overwhelming probability.

So if somebody puts a 0 in here, they must be using some weird algorithm, and use of this weird algorithm will be detectable by 3rd parties by directly observing the commitments (and grinding through possible values). So sidechannel leaks would be the least of their concerns.

Fixing this "properly" would require adding functionality upstream in libsecp-zkp or even libsecp, which does not seem worthwhile for this edge case.